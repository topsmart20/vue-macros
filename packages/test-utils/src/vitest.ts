import path from 'node:path'
import glob, { type Options as GlobOptions } from 'fast-glob'
import { describe, expect, test } from 'vitest'
import { normalizePath } from '@rollup/pluginutils'

interface Options {
  params?: [name: string, values?: any[]][]
  promise?: boolean
}

const SKIP_VUE2 = !!process.env.SKIP_VUE2

export async function testFixtures(
  globs: string | string[],
  exec: (args: Record<string, any>, id: string) => any,
  options?: GlobOptions & Options,
): Promise<void>
export async function testFixtures(
  files: Record<string, string>,
  exec: (args: Record<string, any>, id: string, code: string) => any,
  options?: Options,
): Promise<void>
export async function testFixtures(
  globsOrFiles: string | string[] | Record<string, string>,
  cb: (args: Record<string, any>, id: string, code: string) => any,
  { params, promise, ...globOptions }: GlobOptions & Options = {},
) {
  let files: Record<string, string | undefined>

  if (typeof globsOrFiles === 'string' || Array.isArray(globsOrFiles)) {
    files = Object.fromEntries(
      (await glob(globsOrFiles, globOptions)).map((file) => [file, undefined]),
    )
  } else {
    files = globsOrFiles
  }

  for (const [id, code] of Object.entries(files)) {
    makeTests(id, code, [[normalizePath(id)], ...(params || [])])()
  }

  function getName(name: string, value: any) {
    return value !== undefined ? `${name} = ${String(value)}` : name
  }

  function makeTests(
    id: string,
    code: string | undefined,
    params: NonNullable<Options['params']>,
    args: Record<string, any> = {},
  ) {
    const [currParams, ...restParams] = params
    const [name, values = [undefined]] = currParams
    if (restParams.length > 0) {
      return () => {
        for (const value of values) {
          const currArgs = { ...args, [name]: value }

          describe(
            getName(name, value),
            makeTests(id, code, restParams, currArgs),
          )
        }
      }
    } else {
      return () => {
        for (const value of values) {
          const testName = getName(name, value)
          const isSkip = testName.includes('vue2') && SKIP_VUE2

          test.skipIf(isSkip)(testName, async () => {
            const currArgs = { ...args, [name]: value }
            const execute = () =>
              cb(currArgs, path.resolve(globOptions.cwd || '/', id), code!)
            if (id.includes('error')) {
              if (promise) {
                await expect(execute()).rejects.toThrowErrorMatchingSnapshot()
              } else {
                expect(execute).toThrowErrorMatchingSnapshot()
              }
            } else if (promise) {
              await expect(
                (execute() as Promise<any>).catch((error) => {
                  console.warn(error)
                  return Promise.reject(error)
                }),
              ).resolves.toMatchSnapshot()
            } else {
              expect(execute()).toMatchSnapshot()
            }
          })
        }
      }
    }
  }
}
