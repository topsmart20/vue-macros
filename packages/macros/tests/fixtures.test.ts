import { resolve } from 'node:path'
import { describe } from 'vitest'
import {
  RollupEsbuildPlugin,
  RollupVue,
  RollupVueJsx,
  rollupBuild,
  testFixtures,
} from '@vue-macros/test-utils'
import VueMacros from '../src/rollup'

describe('fixtures', async () => {
  await testFixtures(
    ['tests/fixtures/**/*.{vue,js,jsx,ts,tsx}', '!tests/fixtures/types.ts'],
    (args, id) => {
      const version = id.includes('vue2') ? 2 : 3
      return rollupBuild(id, [
        VueMacros({
          setupSFC: true,
          version,
          plugins: {
            vue: RollupVue({
              include: [/\.vue$/, /\.setup\.[cm]?[jt]sx?/],
            }),
            vueJsx: RollupVueJsx(),
          },
        }),
        RollupEsbuildPlugin({
          target: 'esnext',
        }),
      ])
    },
    {
      cwd: resolve(__dirname, '..'),
      promise: true,
    }
  )
})
