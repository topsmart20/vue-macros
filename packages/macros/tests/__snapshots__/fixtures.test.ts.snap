// Vitest Snapshot v1

exports[`fixtures > tests/fixtures/basic.vue 1`] = `
"import { defineComponent } from 'vue';

var __emitHelper = (emitFn, key, value, ...args) => {
  emitFn(key, value);
  return args.length > 0 ? args[0] : value
};

const name = \\"Foo\\";
const __default__ = defineComponent({
  name
});
var _sfc_main = /* @__PURE__ */ defineComponent({
  ...__default__,
  props: {
    modelValue: { type: String, required: true },
    title: { type: String, required: true }
  },
  emits: [\\"update:modelValue\\", \\"change\\"],
  setup(__props, { emit }) {
    const { modelValue, title } = __props;
    const handleClick = () => {
      emit(\\"change\\");
      __emitHelper(emit, \\"update:modelValue\\", \\"hello, \\" + title);
    };
    return () => {
    };
  }
});

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

var basic = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { basic as default };
"
`;

exports[`fixtures > tests/fixtures/hoist-static.vue 1`] = `
"import { defineComponent } from 'vue';

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

const name = 'App';
const propName = 'title';
const __default__ = defineComponent({
  name,
});
const _sfc_main = /*#__PURE__*/Object.assign(__default__, {
  props: [propName],
  setup(__props) {







return () => {}
}

});
var hoistStatic = /*#__PURE__*/_export_sfc(_sfc_main, [__FILE__]);

export { hoistStatic as default };
"
`;

exports[`fixtures > tests/fixtures/hoist-static-scope.vue 1`] = `
"import { defineComponent, openBlock, createElementBlock } from 'vue';

const [name] = [\\"VarCompName\\"];
const __default__ = defineComponent({
  name
});
var _sfc_main = /* @__PURE__ */ defineComponent({
  ...__default__,
  setup(__props) {
    const test = (name2) => name2;
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock(\\"div\\", null, \\"1\\");
    };
  }
});

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

var hoistStaticScope = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { hoistStaticScope as default };
"
`;

exports[`fixtures > tests/fixtures/vue2-model.vue 1`] = `
"import { defineComponent } from 'vue';

var __emitHelper = (emitFn, key, value, ...args) => {
  emitFn(key, value);
  return args.length > 0 ? args[0] : value
};

const __default__ = defineComponent({
  name: \\"Foo\\",
  model: {
    prop: \\"value\\",
    event: \\"changeInput\\"
  }
});
var _sfc_main = /* @__PURE__ */ defineComponent({
  ...__default__,
  props: {
    modelValue: { type: String, required: true },
    value: { type: String, required: true },
    title: { type: String, required: true }
  },
  emits: [\\"update:modelValue\\", \\"changeInput\\", \\"change\\"],
  setup(__props, { emit }) {
    const { modelValue, value, title } = __props;
    const handleClick = () => {
      emit(\\"change\\");
      __emitHelper(emit, \\"update:modelValue\\", \\"hello, \\" + title);
      __emitHelper(emit, \\"changeInput\\", \\"Word,\\" + title);
    };
    return () => {
    };
  }
});

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

var vue2Model = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { vue2Model as default };
"
`;
